### PR TITLE
[SILOptimizer] Add pipeline execution request

### PR DESCRIPTION
**Pull request by hamishknight**
Add `ExecuteSILPipelineRequest` which executes a pipeline plan on a given SIL (and possibly IRGen) module. This serves as a top-level request for the SILOptimizer that we'll be able to hang dependencies off.

### CODE REVIEW

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Review by CodaFi — APPROVED** [review]
Requests LGTM

**Comment by hamishknight** [comment]
@gottesmm Mind also taking a look?

**Comment by hamishknight** [comment]
Ping @gottesmm

File: include/swift/AST/ASTContext.h [hunk 1]
```diff
@@ -962,6 +963,15 @@ class ASTContext final {
   /// Each kind and SourceFile has its own cache for a Type.
   Type &getDefaultTypeRequestCache(SourceFile *, KnownProtocolKind);
 
+  using SILTransformCtors = ArrayRef<SILTransform *(*)(void)>;
```
Comment (gottesmm) [inline include/swift/AST/ASTContext.h]: SILTransformCtors => SILTransformConstructor?

File: include/swift/SIL/SILModule.h [hunk 2]
```diff
@@ -720,6 +720,9 @@ inline llvm::raw_ostream &operator<<(llvm::raw_ostream &OS, const SILModule &M){
   return OS;
 }
 
+void simple_display(llvm::raw_ostream &out, const SILModule *M);
+SourceLoc extractNearestSourceLoc(const SILModule *SM);
```
Comment (gottesmm) [inline include/swift/SIL/SILModule.h]: These need comments. Tell me why I would want to use this and what for?

**Review by gottesmm — COMMENTED** [review]
I want to look in more detail here. But can you split the IRGenTransform change into a separate commit from the request change? This patch is really those 2 things mushed together. Commits should be clear and focused, not muddled.

File: include/swift/AST/ASTContext.h (hunk 1 above)
Comment (hamishknight) [reply include/swift/AST/ASTContext.h]: @gottesmm Hmm, I'd prefer to keep the name plural as it's referring to an array – `SILTransformConstructors` perhaps?

File: include/swift/SIL/SILModule.h (hunk 2 above)
Comment (hamishknight) [reply include/swift/SIL/SILModule.h]: @gottesmm Sure thing! Though they shouldn't be called directly, they should only really be called by the evaluator's internals.

**Comment by hamishknight** [comment]
@gottesmm Thanks for the review!

> But can you split the IRGenTransform change into a separate commit from the request change? 

Could you elaborate on what you mean by this? As far as I can tell, the IRGen transform specific stuff is contained in https://github.com/apple/swift/pull/29552/commits/71bc8d4b44ccf4332377823bf39b21a8c74aff7f.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test Linux platform

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Comment by hamishknight** [comment]
Ping @gottesmm

**Comment by gottesmm** [comment]
I don't have any problems with this in principle. My only question is for @atrick. I know that we have run into problems around the pass manager and function/module pass managers. We may want that abstraction here in this API since you are creating pass managers yourself. Not completely sure.

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test

**Review by atrick — COMMENTED** [review]
Looks like this should work just fine.

**Comment by hamishknight** [comment]
https://github.com/apple/llvm-project/pull/695

@swift-ci please test macOS platform

**Pull request merged**